leakpro/attacks/mia_attacks/yoqo.py

```diff
@@ -0,0 +1,340 @@
+
+            # Filter the audit data
+            self.audit_data_indices = self.audit_dataset["data"][mask]
+            self.in_indices_masks = self.in_indices_masks[mask, :]
+
+            # Filter IN and OUT members
+            self.in_members = np.arange(np.sum(mask[self.audit_dataset["in_members"]]))
+            num_out_members = np.sum(mask[self.audit_dataset["out_members"]])
+            self.out_members = np.arange(len(self.in_members), len(self.in_members) + num_out_members)
+
+            assert len(self.audit_data_indices) == len(self.in_members) + len(self.out_members)
+
+            if len(self.audit_data_indices) == 0:
+                raise ValueError("No points in the audit dataset are used for the shadow models")
+
+        else:
+            self.audit_data_indices = self.audit_dataset["data"]
+            self.in_members = self.audit_dataset["in_members"]
+            self.out_members = self.audit_dataset["out_members"]
+
+        # Check offline attack for possible IN- sample(s)
+        if not self.online:
+            count_in_samples = np.count_nonzero(self.in_indices_masks)
+            if count_in_samples > 0:
+                logger.info(f"Some shadow model(s) contains {count_in_samples} IN samples in total for the model(s)")
+                logger.info("This is not an offline attack!")
+
+        # Select a subset of the data to audit
+        if self.n_audits < 0:
+            self.n_audits = len(self.audit_data_indices)
+        else:
+            self.in_members = self.in_members[:(self.n_audits//2)]
+            self.out_members = self.out_members[:(self.n_audits - len(self.in_members))]
+            self.audit_data_indices = np.concatenate(
+                (
+                    self.audit_data_indices[self.in_members],
+                    self.audit_data_indices[self.out_members]
+                )
+            )
+            self.in_indices_masks = np.concatenate(
+                (
+                    self.in_indices_masks[self.in_members,:],
+                    self.in_indices_masks[self.out_members,:]
+                ),
+                axis = 0
+            )
+
+    def _prepare_optimization_objective(self:Self) -> None:
+        if self.online:
+            if self.output_shape == 1:
+                self._optimization_objective = self._optimization_objective_online_binary
+            else:
+                self._optimization_objective = self._optimization_objective_online
+        else:
+            self.mse = MSELoss()
+
+            if self.output_shape == 1:
+                self._optimization_objective = self._optimization_objective_offline_binary
+            else:
+                self._optimization_objective = self._optimization_objective_offline
+
+        self.weights = 1 + (self.alpha - 1) * self.in_indices_masks
+        self.weights = Tensor(self.weights)
+        self.batch_size = 5000
```

Comment:
This may result in a very large memory allocation. Perhaps stick with default value in `get_dataloader()`?